File: src/main/java/com/mapzen/valhalla/Instruction.java
package com.mapzen.valhalla;

import com.mapzen.helpers.DistanceFormatter;
import com.mapzen.ontheroad.R;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;
import android.location.Location;
import android.util.Log;

import java.util.Locale;

import static com.mapzen.osrm.Route.SNAP_PROVIDER;

public class Instruction {
    public static final int NO_TURN = 0;
    public static final int GO_STRAIGHT = 1;
    public static final int TURN_SLIGHT_RIGHT = 2;
    public static final int TURN_RIGHT = 3;
    public static final int TURN_SHARP_RIGHT = 4;
    public static final int U_TURN = 5;
    public static final int TURN_SHARP_LEFT = 6;
    public static final int TURN_LEFT = 7;
    public static final int TURN_SLIGHT_LEFT = 8;
    public static final int REACH_VIA_POINT = 9;
    public static final int HEAD_ON = 10;
    public static final int ENTER_ROUND_ABOUT = 11;
    public static final int LEAVE_ROUND_ABOUT = 12;
    public static final int STAY_ON_ROUND_ABOUT = 13;
    public static final int START_AT_END_OF_STREET = 14;
    public static final int YOU_HAVE_ARRIVED = 15;
    public static final int ENTER_AGAINST_ALLOWED_DIRECTION = 16;
    public static final int LEAVE_AGAINST_ALLOWED_DIRECTION = 17;
    public static final int INSTRUCTION_COUNT = 18;

    private JSONArray json;
    private int turn, distanceInMeters;
    private Location location = new Location(SNAP_PROVIDER);
    private int liveDistanceToNext = -1;

    public int getLiveDistanceToNext() {
        return liveDistanceToNext;
    }

    public void setLiveDistanceToNext(int liveDistanceToNext) {
        this.liveDistanceToNext = liveDistanceToNext;
    }

    public Instruction(JSONArray json) throws JSONException {
        if (json.length() < 8) {
            throw new JSONException("too few arguments");
        }
        this.json = json;
        setTurnInstruction(parseTurnInstruction(json));
        setDistance(json.getInt(2));
    }

    /**
     * Used for testing. Do not remove.
     */
    @SuppressWarnings("unused")
    protected Instruction() {
    }

    public int getTurnInstruction() {
        return turn;
    }

    public void setTurnInstruction(int turn) {
        this.turn = turn;
    }

    public String getHumanTurnInstruction(Context context) {
        switch (turn) {
            case NO_TURN:
                return context.getString(R.string.no_turn);
            case GO_STRAIGHT:
                return context.getString(R.string.go_straight);
            case TURN_SLIGHT_RIGHT:
                return context.getString(R.string.turn_slight_right);
            case TURN_RIGHT:
                return context.getString(R.string.turn_right);
            case TURN_SHARP_RIGHT:
                return context.getString(R.string.turn_sharp_right);
            case U_TURN:
                return context.getString(R.string.u_turn);
            case TURN_SHARP_LEFT:
                return context.getString(R.string.turn_sharp_left);
            case TURN_LEFT:
                return context.getString(R.string.turn_left);
            case TURN_SLIGHT_LEFT:
                return context.getString(R.string.turn_slight_left);
            case REACH_VIA_POINT:
                return context.getString(R.string.reach_via_point);
            case HEAD_ON:
                return context.getString(R.string.head_on);
            case ENTER_ROUND_ABOUT:
                return context.getString(R.string.enter_round_about);
            case LEAVE_ROUND_ABOUT:
                return context.getString(R.string.leave_round_about);
            case STAY_ON_ROUND_ABOUT:
                return context.getString(R.string.stay_on_round_about);
            case START_AT_END_OF_STREET:
                return context.getString(R.string.start_at_end_of_street);
            case YOU_HAVE_ARRIVED:
                return context.getString(R.string.you_have_arrived);
            case ENTER_AGAINST_ALLOWED_DIRECTION:
                return context.getString(R.string.enter_against_allowed_direction);
            case LEAVE_AGAINST_ALLOWED_DIRECTION:
                return context.getString(R.string.leave_against_allowed_direction);
            default:
                return context.getString(R.string.no_turn);
        }
    }

    public boolean skip() throws JSONException {
        String raw = json.getString(1);
        return raw.startsWith("{") && raw.endsWith("}");
    }

    public String getName() throws JSONException {
        String raw = json.getString(1);
        if (raw.startsWith("{") && raw.endsWith("}")) {
            JSONObject nameObject = new JSONObject(raw);
            return nameObject.getString("highway");
        } else {
            return raw;
        }
    }

    public int getDistance() {
        return distanceInMeters;
    }

    public void setDistance(int distanceInMeters) {
        this.distanceInMeters = distanceInMeters;
    }

    public String getFormattedDistance() {
        return DistanceFormatter.format(distanceInMeters);
    }

    public String getDirection() throws JSONException {
        return json.getString(6);
    }

    public int getPolygonIndex() throws JSONException {
        return json.getInt(3);
    }

    public float getDirectionAngle() throws JSONException {
        String direction = getDirection();
        float angle = 0;
        if (direction.equals("NE")) {
            angle = 315.0f;
        } else if (direction.equals("E")) {
            angle = 270.0f;
        } else if (direction.equals("SE")) {
            angle = 225.0f;
        } else if (direction.equals("S")) {
            angle = 180.0f;
        } else if (direction.equals("SW")) {
            angle = 135.0f;
        } else if (direction.equals("W")) {
            angle = 90.0f;
        } else if (direction.equals("NW")) {
            angle = 45.0f;
        }
        return angle;
    }

    public int getRotationBearing() throws JSONException {
        return 360 - json.getInt(7);
    }

    public int getBearing() throws JSONException {
        return json.getInt(7);
    }

    public Location getLocation() {
        return location;
    }

    public void setLocation(Location location) {
        this.location = location;
    }

    public String getFullInstruction(Context context) throws JSONException {
        return getFullInstructionBeforeAction(context);
    }

    public String getFullInstructionBeforeAction(Context context) throws JSONException {
        if (turn == HEAD_ON || turn == GO_STRAIGHT) {
            return context.getString(R.string.full_instruction_before_straight,
                    getHumanTurnInstruction(context), getName(),
                    DistanceFormatter.format(distanceInMeters, true));
        } else if (turn ==YOU_HAVE_ARRIVED) {
            return context.getString(R.string.full_instruction_destination,
                    getHumanTurnInstruction(context), getName());
        } else {
            return context.getString(R.string.full_instruction_before_default,
                    getHumanTurnInstruction(context), getName(),
                    DistanceFormatter.format(distanceInMeters, true));
        }
    }

    public String getFullInstructionAfterAction(Context context) throws JSONException {
        if (turn == YOU_HAVE_ARRIVED) {
            return context.getString(R.string.full_instruction_destination,
                    getHumanTurnInstruction(context), getName());
        }

        return context.getString(R.string.full_instruction_after, getName(),
                DistanceFormatter.format(distanceInMeters, false));
    }

    public String getSimpleInstruction(Context context) throws JSONException {
        return context.getString(R.string.simple_instruction, getHumanTurnInstruction(context),
                getName());
    }

    @Override
    public String toString() {
        try {
            return String.format(Locale.US, "Instruction: (%.5f, %.5f) %s %s LiveDistanceTo: %d",
                    location.getLatitude(), location.getLongitude(), turn,
                    getName(), liveDistanceToNext);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public boolean equals(Object obj) {
        if (getClass() != obj.getClass()) {
            return false;
        }
        Instruction other = (Instruction) obj;
        try {
            return (getTurnInstruction() == other.getTurnInstruction()
                    && getBearing() == other.getBearing()
                    && getLocation().getLatitude() == other.getLocation().getLatitude()
                    && getLocation().getLongitude() == other.getLocation().getLongitude());
        } catch (JSONException e) {
            Log.e("Json exception", "Unable to get bearing", e);
            return false;
        }
    }

    private int parseTurnInstruction(JSONArray json) throws JSONException {
        String turn = json.getString(0);
        String[] split = turn.split("-");
        return Integer.valueOf(split[0]);
    }

    public String getSimpleInstructionAfterAction(Context context) throws JSONException {
        if (turn == YOU_HAVE_ARRIVED) {
            return getFullInstructionBeforeAction(context);
        }

        return context.getString(R.string.simple_instruction_after, getName());
    }
}

File: src/main/java/com/mapzen/valhalla/Instruction.kt
package com.mapzen.valhalla

import com.mapzen.helpers.DistanceFormatter
import com.mapzen.model.Location
import org.json.JSONException
import org.json.JSONObject
import java.util.Locale

open class Instruction {
    companion object {
        val KM_TO_METERS = 1000
        val MI_TO_METERS = 1609.344
    }

    val NONE : Int = 0
    val START : Int = 1
    val START_RIGHT : Int = 2
    val START_LEFT : Int = 3
    val DESTINATION : Int = 4
    val DESTINATION_RIGHT : Int = 5
    val DESTINATION_LEFT : Int = 6
    val BECOMES : Int = 7
    val CONTINUE : Int = 8
    val SLIGHT_RIGHT : Int = 9
    val RIGHT : Int = 10
    val SHARP_RIGHT : Int = 11
    val U_TURN_RIGHT: Int = 12
    val U_TURN_LEFT : Int = 13
    val SHARP_LEFT : Int = 14
    val LEFT : Int = 15
    val SLIGHT_LEFT : Int = 16
    val RAMP_STRAIGHT : Int = 17
    val RAMP_RIGHT : Int = 18
    val RAMP_LEFT : Int = 19
    val EXIT_RIGHT : Int = 20
    val EXIT_LEFT : Int = 21
    val STAY_STRAIGHT : Int = 22
    val STAY_RIGHT : Int = 23
    val STAY_LEFT : Int = 24
    val MERGE : Int = 25
    val ROUNDABOUT_ENTER : Int = 26
    val ROUNDABOUT_EXIT: Int = 27
    val FERRY_ENTER: Int = 28
    val FERRY_EXIT: Int = 29

    private var json: JSONObject? = null;

    var turnInstruction: Int = 0
    var distance: Int = 0
    var location: Location = Location()
    var liveDistanceToNext: Int = -1
    var bearing: Int = 0

    constructor(json: JSONObject) : this(json, Router.DistanceUnits.KILOMETERS) { }

    constructor(json: JSONObject, units: Router.DistanceUnits) {
        if (json.length() < 6) {
            throw JSONException("too few arguments")
        }
        this.json = json
        turnInstruction = parseTurnInstruction(json)

        val raw = json.getDouble("length")
        when (units) {
            Router.DistanceUnits.KILOMETERS -> distance = Math.round(raw * KM_TO_METERS).toInt()
            Router.DistanceUnits.MILES -> distance = Math.round(raw * MI_TO_METERS).toInt()
        }
    }

    /**
     * Used for testing. Do not remove.
     */
    @SuppressWarnings("unused")
    protected constructor() {
    }

    fun getIntegerInstruction(): Int {
       return turnInstruction;
    }

    fun getHumanTurnInstruction(): String? {
        return json?.getString("instruction");
    }

    fun skip(): Boolean {
        if(json!!.optJSONArray("street_names") == null && json!!.getInt("type") != DESTINATION ) {
            return true
        }
        return false;
    }

    fun getBeginStreetNames(): String {
        if (json?.has("begin_street_names") ?: false) {
            var streetName = "";
            val numStreetNames = (json!!.getJSONArray("begin_street_names").length())
            for(i in 0..numStreetNames - 1) {
                streetName += json!!.getJSONArray("begin_street_names").get(i);
                if((numStreetNames > 1) && (i < numStreetNames - 1)) {
                    streetName += "/"
                }
            }
            return streetName;
        }

        return ""
    }


    fun getName(): String {
        if (json?.has("street_names") ?: false) {
            var streetName = "";
            val numStreetNames = (json!!.getJSONArray("street_names").length())
            for(i in 0..numStreetNames - 1) {
                streetName += json!!.getJSONArray("street_names").get(i);
                if((numStreetNames > 1) && (i < numStreetNames - 1)) {
                    streetName += "/"
                }
            }
            return streetName;
        }

        return json?.getString("instruction") ?: ""
    }

    fun getFormattedDistance(): String {
        return DistanceFormatter.format(distance.toInt())
    }

    fun getBeginPolygonIndex(): Int {
        return json!!.getInt("begin_shape_index");
    }

    fun getEndPolygonIndex(): Int {
        return json!!.getInt("end_shape_index");
    }

    fun getDirectionAngle(): Float {
        val direction = bearing
        var angle: Float = 0.0f
        if (direction >= 315.0 && direction <= 360.0 ) {
            angle = 315.0.toFloat()
        } else if (direction >= 270.0 && direction < 315.0) {
            angle = 270.0.toFloat()
        } else if (direction >= 225.0 && direction < 270.0) {
            angle = 225.0.toFloat()
        } else if (direction >= 180.0 && direction < 225.0) {
            angle = 180.0.toFloat()
        } else if (direction >= 135.0 && direction < 180.0) {
            angle = 135.0.toFloat()
        } else if (direction >= 90.0 && direction < 135.0) {
            angle = 90.0.toFloat()
        } else if (direction >= 45.0 && direction < 90.0) {
            angle = 45.0.toFloat()
        } else if (direction >= 0.0 && direction < 45.0) {
            angle = 0.0.toFloat()
        }
        return angle
    }

    fun getDirection(): String {
        var direction = ""
        if (bearing >= 315.0 && bearing >= 360.0 ) {
            direction = "NE"
        } else if (bearing >= 270.0 && bearing < 315.0) {
            direction = "E"
        } else if (bearing >= 225.0 && bearing < 270.0) {
            direction = "SE"
        } else if (bearing >= 180.0 && bearing < 225.0) {
            direction = "S"
        } else if (bearing >= 135.0 && bearing < 180.0) {
            direction = "SW"
        } else if (bearing >= 90.0 && bearing < 135.0) {
            direction = "W"
        } else if (bearing >= 45.0 && bearing < 90.0) {
            direction = "NW"
        } else if (bearing >= 0.0 && bearing < 45.0) {
            direction = "N"
        }
        return direction
    }

    fun getRotationBearing(): Int {
        return 360 - bearing
    }

    override fun toString(): String {
        var name = ""
        try {
            name = getName()
        } catch (e: JSONException) {
            System.out.println("Json exception unable to get name" + e.stackTrace)
        }

        return java.lang.String.format(Locale.US, "Instruction: (%.5f, %.5f) %s %s" +
                "LiveDistanceTo: %d", location.latitude, location.longitude,
                turnInstruction, name, liveDistanceToNext)
    }

    override fun equals(obj: Any?): Boolean {
        if (obj == null) {
            return false
        }

        if (javaClass != obj.javaClass) {
            return false
        }

        val other = obj as Instruction
        return (turnInstruction == other.turnInstruction
                && bearing == other.bearing
                && location.latitude == other.location.latitude
                && location.longitude == other.location.longitude)
    }

    private fun parseTurnInstruction(json: JSONObject): Int =
            json.getInt("type")

    fun getVerbalPreTransitionInstruction(): String {
        if (json?.has("verbal_pre_transition_instruction") ?: false) {
            return json?.getString("verbal_pre_transition_instruction") ?: ""
        }

        return ""
    }

    fun getVerbalTransitionAlertInstruction(): String {
        if (json?.has("verbal_transition_alert_instruction") ?: false) {
            return json?.getString("verbal_transition_alert_instruction") ?: ""
        }

        return ""
    }

    fun getVerbalPostTransitionInstruction(): String {
        if (json?.has("verbal_post_transition_instruction") ?: false) {
            return json?.getString("verbal_post_transition_instruction") ?: ""
        }

        return ""
    }
}
